functions/src/index.ts

```diff
@@ -166,3 +177,141 @@ export const loadFirestore = functions.https.onRequest(
     response.status(200).send();
   }
 );
+
+/*
+ * Listen for changes to the User document
+ * and initiate the appropriate Glean ping(s)
+ */
+export const handleUserChangesImpl = async function (
+  change: Change<DocumentSnapshot>,
+  context: EventContext
+): Promise<boolean> {
+  const userID = context.params.userID;
+  const rallyID = await getRallyIdForUser(userID);
+  if (!rallyID) {
+    // Without Rally ID, we can't make any Glean pings
+    // This is bad and should be flagged for inspection
+    throw new Error(
+      `Unable to obtain Rally ID for user ID ${userID}. Aborting Glean ping process.`
+    );
+  }
+
+  // Get an object with the current document value.
+  // If the document does not exist, it has been deleted.
+  const newUser = change.after.exists ? change.after.data() : null;
+
+  // Get the old document, to compare the enrollment state.
+  const oldUser = change.before.exists ? change.before.data() : null;
+
+  if (!newUser || (oldUser && oldUser.enrolled === true && !newUser.enrolled)) {
+    // User document has been deleted
+    functions.logger.info(`Sending unenrollment ping for user ID ${userID}`);
+    await gleanPings.platformUnenrollment(rallyID);
+    return true;
+  }
+
+  if ((!oldUser || !oldUser.enrolled) && newUser.enrolled === true) {
+    // User just enrolled
+    functions.logger.info(`Sending enrollment ping for user ID ${userID}`);
+    await gleanPings.platformEnrollment(rallyID);
+  }
+
+  if (
+    ((!oldUser || !oldUser.demographicsData) && newUser.demographicsData) ||
+    (oldUser && oldUser.demographicsData && !newUser.demographicsData) ||
+    (oldUser &&
+      newUser &&
+      !isDeepStrictEqual(oldUser.demographicsData, newUser.demographicsData))
+  ) {
+    // User updated demographicsData
+    functions.logger.info(`Sending demographics ping for user ID ${userID}`);
+    await gleanPings.demographics(rallyID, newUser.demographicsData);
+  }
+
+  return true;
+};
+
+export const handleUserChanges = functions.firestore
+  .document("users/{userID}")
+  .onWrite(handleUserChangesImpl);
+
+/*
+ * Listen for changes to the Study document
+ * and initiate the appropriate Glean ping(s)
+ */
+export const handleUserStudyChangesImpl = async function (
+  change: Change<DocumentSnapshot>,
+  context: EventContext
+): Promise<boolean> {
+  const userID = context.params.userID;
+  const firebaseStudyID = context.params.studyID;
+  const rallyID = await getRallyIdForUser(userID);
+  if (!rallyID) {
+    // Without Rally ID, we can't make any Glean pings
+    // This is bad and should be flagged for inspection
+    throw new Error(
+      `Unable to obtain Rally ID for user ID ${userID}. Aborting Glean ping process.`
+    );
```

Comment:
This exception is currently throwing when I try to delete a user (it looks like the website needs to call this from "Leave Rally" on the website):
```
>  {"severity":"WARNING","message":"Snapshot has no readTime. Using now()"}
⚠  functions: Error: Unable to obtain Rally ID for user ID jg7G6Q5vtdkOzFq2KLjqhnZgLXa2. Aborting Glean ping process.
    at handleUserChangesImpl (file:///Users/mozilla/src/rally-web-platform/functions/lib/index.js:149:15)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async runFunction (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:533:9)
    at async runBackground (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:545:5)
    at async processBackground (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:528:5)
    at async invokeTrigger (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:610:13)
    at async handleMessage (/Users/mozilla/src/rally-web-platform/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:702:9)
⚠  Your function was killed because it raised an unhandled error.
```

I used the Firebase emulator to delete the user for now, maybe the trigger that deletes the user's docs in Firestore is running first?

Reply:
Oh I think what you want here is to watch for the `extensionUserDoc`, so you can get the "old" one when it gets deleted, since that's the document where the `rallyId` is stored. I had started suggesting a commit here, but I realized I was telling you to listen on the wrong document :)

Reply:
@rhelmer is it ok if I cheat (read: simplify things 😛) and just wait to delete the `extensionUserDoc` in the user deletion logic from `handleUserChangesImpl`, after I've obtained the rallyId via the normal route?

Reply:
Yeah I think that's fine as long as it is deterministic etc.